### PR TITLE
Runtime events counters fixes

### PR DESCRIPTION
**Pull request by sadiqj**
This is hopefully the last mop-up of runtime events fixes that have come out of wider testing. These are specifically fixes for the counters.

* https://github.com/ocaml/ocaml/pull/11400/commits/cd1537c7fea8d6fe865163a1ec9565ca25db0ee8 fixes a missing counter in the OCaml API but that is present in the implementation (without it later counters are out of sync)
* https://github.com/ocaml/ocaml/pull/11400/commits/5dd11c1af5a895b911e1e9f17e9ff06d49cc6705 removes counters that had probes in 4.x eventlog but not in multicore's eventlog or now runtime events. We should think about which ones would be useful to have back and re-add them after the mark-prefetching changes land.
* https://github.com/ocaml/ocaml/pull/11400/commits/993c0e2a7d8812b82a17e49422cad51f47e07623 there was a minor heap size change event but no matching probe.
* https://github.com/ocaml/ocaml/pull/11400/commits/5bc49c394d174e3e8093f7622dd88e1e218675f5 similar to above but for major adjust gc speed.
* ~~https://github.com/ocaml/ocaml/commit/614a9373bbd595ca51f88ebd1958f5c2a111d7d4 these are some new counter events and probes that return the per-domain major heap size/blocks/pool live words and a few other stats that runtime keeps anyway. Given these are new events/probes I'm happy to move this to a separate PR if it simplifies things. I may need some input from @gasche on the gc stats side. I can't quite satisfy myself that we're not leaving out the orphaned pools.~~

This will be of interest to @damiendoligez and @kayceesrk

### CODE REVIEW

**Comment by gasche** [comment]
> Given these are new events/probes I'm happy to move this to a separate PR if it simplifies things

No strong opinion, but I think a bit more discussion will be needed for those.

>  I may need some input from @gasche on the gc stats side. I can't quite satisfy myself that we're not leaving out the orphaned pools.

### Orphaned heap

My understanding is that:
- each domain logically owns a part of the major heap
- and there is some "orphaned heap" that has no domain owner and is gradually adopted over time (there are in fact several different orphaning processes, for example "orphan work" from finalisers or ephemerons, and "orphan free space" in the major heap)

At a given point in time, if you want to report on all resources, you need to list the local heap of each domain, and the orphan heap. But here we can choose when it is that we report these resources, so we would also have the option of reporting them at a time where we know the orphan heap is empty. (I *think* that it is the case after the call to `caml_cycle_heap` within `cycle_all_domains_callback`.)

### How to report domain-local counters

I think there is a bug in the current version of the commit, where you report domain-local heap counters within the `barrier_is_final` conditional, that is, in the part of the STW section that is only executed by a single domain. So a single domain reports its domain-local statistics at each cycle, this is wrong.

We probably want one of:
1. Each domain reports its domain-local statistics (so this code should be outside the exclusive part of the barrier, and it could probably go outside the barrier completely). This is assuming that consumers will be able to distinguish which counter comes from which domain, and sum them up to get a global view of the heap. Is there enough information in the runtime events to distinguish different domains writing their domain-local view of the same counter?
2. Or we only publish global statistics by summing up the counters in the runtime and publishing the aggregate. Then this code should probably cooperate with the gc_stats logic / be placed in those stats functions.

(1 sounds better because it gives more power to the user to understand what's going on, but it also requires more sophistication on the consumer end, do we have an example of consumer correctly doing this?)

### When to report domain-local counters?

You made the choice to report on domain-local heap statistics within the `cycle_all_domains` logic, which I vaguely understand corresponds to a global stop-the-world synchronization phase at the end of each full major GC cycle. Is this the right time to report on these statistics? Maybe it's not often enough, we want more information during the cycle? I don't know, I think that people familiar with the major GC would know.

### What to report?

I looked at the sequential runtime for guidance, but there are basically no global heap statistics in Eventlog in the sequential runtime. I think that the CAML_INSTR stuff grew organically out of the need to get finer-grained metrics than what the "gc stats" provide, complementary to those gc stats -- extra stuff for people who already ask for the stats.

For eventlog and now runtime-events the situation is a bit different: we do not assume that users observe the runtime-events output coupled with the GC stats, they look at one or the other but cannot easily see both in the same monitoring setup. So it makes sense to include more stats as you are doing, and in fact we may want to guarantee systematically the runtime-events output subsume the gc stats. This suggests, again, that the two may benefit from being more tightly coupled together.

**Comment by sadiqj** [comment]
Thanks for taking a look Gabriel

> I think there is a bug in the current version of the commit, where you report domain-local heap counters within the `barrier_is_final` conditional, that is, in the part of the STW section that is only executed by a single domain. So a single domain reports its domain-local statistics at each cycle, this is wrong.

Yes, good catch. I missed the `barrier_is_final` so you're right that this would only happen on one domain per major cycle.

I _think_ if we go for straight after https://github.com/ocaml/ocaml/blob/614a9373bbd595ca51f88ebd1958f5c2a111d7d4/runtime/major_gc.c#L1046 then all global (orphaned) pools must have been adopted, along with their stats. That _should_ mean we only need to report domain-local at that point?

>     1. Each domain reports its domain-local statistics (so this code should be outside the exclusive part of the barrier, and it could probably go outside the barrier completely). This is assuming that consumers will be able to distinguish which counter comes from which domain, and sum them up to get a global view of the heap. Is there enough information in the runtime events to distinguish different domains writing their domain-local view of the same counter?

This was my preferred model. A ring is only attached to one domain so you will be able to track the statistics over time for that domain. If you have events when the domain is spawned, you'll also have the PID. (A future feature could be having a separate area in the `.events` file for storing the names/PIDs of the domains rings are attached to).

> (1 sounds better because it gives more power to the user to understand what's going on, but it also requires more sophistication on the consumer end, do we have an example of consumer correctly doing this?)

No. Like most things I suspect users won't actually build consumers directly but rely on tooling that's doing it correctly. Hopefully.

> ### When to report domain-local counters?
> 
It wasn't entirely clear there was a better time to report. I think at this point the `live` values represent everything that was marked or allocated in the previous cycle. It's not entirely clear to me what that means halfway through a cycle.

> ### What to report?

My thinking has been that the instrumented runtime events and probes should be data that's useful for making design decisions with the GC. I know there's some work coming on re-evaluating the allocator and I'm sure there are instrumented probes which would be useful for that.

The standard runtime probes I've been trying to keep to things that you could realistically remedy by changing your program's behaviour or the GC params.

**Comment by gasche** [comment]
> That should mean we only need to report domain-local at that point?

Yes, I think this would be a sensible approach.

> Like most things I suspect users won't actually build consumers directly but rely on tooling that's doing it correctly. Hopefully.

Where does the code for that tooling lies? I think it would be useful to actually build a prototype of matching tooling to make sure that whatever approach you use is actually amenable to providing sensible GC stats.

> I know there's some work coming on re-evaluating the allocator 

I didn't know and am mildly curious, do you have some pointers? (Is it best-fit related?)

**Comment by sadiqj** [comment]
> Yes, I think this would be a sensible approach.

Moved in https://github.com/ocaml/ocaml/pull/11400/commits/69d6bb9e37b77ea758dbef2a0d749c38ed840673

> Where does the code for that tooling lies? I think it would be useful to actually build a prototype of matching tooling to make sure that whatever approach you use is actually amenable to providing sensible GC stats.

@kayceesrk and I have been working on a set of tools in https://github.com/sadiqj/runtime_events_tools , there should be a new opam release soon (it was blocked on getting hdr_histogram packaged)

> > I know there's some work coming on re-evaluating the allocator
> 
> I didn't know and am mildly curious, do you have some pointers? (Is it best-fit related?)

@jonludlam probably has more details on the plans around this.

**Comment by jonludlam** [comment]
> > I know there's some work coming on re-evaluating the allocator
> 
> I didn't know and am mildly curious, do you have some pointers? (Is it best-fit related?)

Yes, best-fit and also compaction. The plan is to do some benchmark/tooling work first in order to be able to evaluate the changes, then look into the implementation. We've not really started in earnest yet so not much to share, but I'm sure we'll be happy to do so when things are underway.

**Comment by sadiqj** [comment]
@gasche @Octachron it would be great to have the fixes from the first four commits in alpha1. Should I remove the final commit that adds the new counters?

**Comment by Octachron** [comment]
Splitting the PR in two and creating a PR for more involved last commit sounds like a very good idea to me.

**Comment by sadiqj** [comment]
Thanks @Octachron . I've removed the final commit, so it's now only fixes in.

File: runtime/gc_ctrl.c [hunk 1]
```diff
@@ -221,6 +221,7 @@ CAMLprim value caml_gc_set(value v)
        (newminwsz != Caml_state->minor_heap_wsz) are both true,
        the current domain reallocates its own minor heap twice. */
     caml_set_minor_heap_size (newminwsz);
+    CAML_EV_COUNTER (EV_C_FORCE_MINOR_SET_MINOR_HEAP_SIZE, 1);
```
Comment (Octachron) [inline runtime/gc_ctrl.c]: Why is this event here rather than inside the body of `caml_set_minor_heap_size`?

Comment (sadiqj) [reply runtime/gc_ctrl.c]: Good call. I've moved it inside `caml_set_minor_heap_size` and actually made sure it's only emitted when there is a forced minor collection: https://github.com/ocaml/ocaml/pull/11400/commits/9da76f9065ea0d11d9b9280d741338df174ffadd

**Review by Octachron — APPROVED** [review]
This simplification of the  runtime event counters looks good to me.

**Comment by sadiqj** [comment]
Thanks @Octachron

**Comment by Octachron** [comment]
The fixes have been cherry-picked to 5.0(f1aea71db567).

**Pull request merged**